src/state.rs

```diff
@@ -591,6 +443,183 @@ impl DemandProxyState {
+
+    // Map of in-progress resolution requests.
+    in_progress: Arc<Mutex<HashMap<String, Arc<Notify>>>>,
+
+    dns_resolver_cfg: ResolverConfig,
+
+    dns_resolver_opts: ResolverOpts,
+}
+
+#[derive(serde::Serialize, Default, Debug, Clone)]
+struct ResolvedDns {
+    hostname: String,
+    ips: HashSet<IpAddr>,
+    #[serde(skip_serializing)]
+    initial_query: Option<std::time::Instant>,
+    // the shortest DNS ttl of all records in the response; used for cache refresh.
+    // we use the shortest ttl rather than just relying on the older records so we don't
+    // load-balance to just the older records as the records with early ttl expire.
+    dns_refresh_rate: std::time::Duration,
+}
+
+impl DnsResolver {
+    fn new(dns_resolver_cfg: ResolverConfig, dns_resolver_opts: ResolverOpts) -> Self {
+        Self {
+            resolved: Arc::new(RwLock::new(HashMap::new())),
+            in_progress: Arc::new(Mutex::new(HashMap::new())),
+            dns_resolver_cfg,
+            dns_resolver_opts,
+        }
+    }
+
+    async fn resolve_host(
+        &self,
+        workload: &Workload,
+        src_workload: &Workload,
+        metrics: Arc<proxy::Metrics>,
+    ) -> Option<ResolvedDns> {
+        let labels = OnDemandDnsLabels::new()
+            .with_destination(workload)
+            .with_source(src_workload);
+        metrics.as_ref().on_demand_dns.get_or_create(&labels).inc();
+
+        // First, check if we've already resolved this host.
+        let hostname = workload.hostname.to_owned();
+        if let Some(resolved) = self._find_resolved_host(&hostname) {
+            return Some(resolved);
+        }
+
+        metrics
+            .as_ref()
+            .on_demand_dns_cache_misses
+            .get_or_create(&labels)
+            .inc();
+
+        // We need to resolve the host. The first request here will create the
+        // notify and will perform the resolution. Requests that follow will
+        // just wait for the results of the first request.
+        let (n, is_first) = self._get_or_create_notify(&hostname);
+        if is_first {
+            // We're the first: perform the resolution of the host.
+            self._resolve_host(workload).await;
+
+            // notify all waiters after the dns resolving task completed
+            n.notify_waiters();
+
+            // All threads that were waiting have been notified and the
+            // local cache has been updated. We can now go ahead and
+            // remove the in-progress notify object.
+            self.in_progress.lock().unwrap().remove(hostname.as_str());
+        } else {
+            // Wait for the in-progress resolution to complete.
+            n.notified().await;
+        }
+
+        // At this point, resolution has completed. Just serve from local
+        // cache.
+        self._find_resolved_host(&hostname)
+    }
+
+    fn _get_or_create_notify(&self, hostname: &String) -> (Arc<Notify>, bool) {
```

Comment:
I think rust has explicit property for function, such as `pub` key word, so I prefer to remove `_` for this private function.